server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java

```diff
@@ -390,8 +390,43 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
      * Format <code>obj</code> using the provided {@link DocValueFormat}.
      * If the format is equals to {@link DocValueFormat#RAW}, the object is returned as is
      * for numbers and a string for {@link BytesRef}s.
+     *
+     * This method will then attempt to parse the formatted value using the specified format,
+     * and throw an IllegalArgumentException if parsing fails.  This in turn prevents us from
+     * returning an after_key which we can't subsequently parse into the original value.
      */
-    static Object formatObject(Object obj, DocValueFormat format) {
+    static Object formatObjectChecked(Object obj, DocValueFormat format) {
+        Object formatted = formatObjectUnchecked(obj, format);
+        if (formatted != null) {
+            Object parsed;
+            // Type Jank ahead
+            if (obj.getClass() == BytesRef.class) {
```

Comment:
Could we avoid this layer of type jank? We already know the type from `formatObjectUnchecked`. Maybe if the checking were performed in a callback or something? I'm not sure that'd be more readable, but maybe one less janky type based switch statement would be worth trying?